### PR TITLE
community[patch]: Support OpenSearch Serverless

### PR DESCRIPTION
**Pull request by huuyafwww**
Fixes
- #2036
- #2302

### CODE REVIEW

File: libs/langchain-community/src/vectorstores/opensearch.ts [hunk 1]
```diff
@@ -27,6 +27,7 @@ interface VectorSearchOptions {
  */
 export interface OpenSearchClientArgs {
   readonly client: Client;
+  readonly service?: 'es' | 'aoss';
```
Comment (huuyafwww) [inline libs/langchain-community/src/vectorstores/opensearch.ts:30]: The service name is adapted from the opensearch-js client library.

https://github.com/opensearch-project/opensearch-js/blob/835ad4ec11e03ec3b0293612226f91b859f2553d/lib/aws/index.d.ts#L22

File: libs/langchain-community/src/vectorstores/opensearch.ts [hunk 2]
```diff
@@ -116,21 +121,34 @@ export class OpenSearchVectorStore extends VectorStore {
...
+          embedding,
+          metadata: documents[idx].metadata,
+          text: documents[idx].pageContent,
+        },
+      ];
+
+      // aoss does not support document id
+      if (this.isAoss) {
+        delete document[0].index?._id;
+      }
```
Comment (huuyafwww) [inline libs/langchain-community/src/vectorstores/opensearch.ts:142]: The following errors that occurred when using OpenSearch Serverless have been addressed
```
Document ID is not supported in create/index operation request
```

File: libs/langchain-community/src/vectorstores/opensearch.ts [hunk 3]
```diff
@@ -116,21 +121,34 @@ export class OpenSearchVectorStore extends VectorStore {
...
+
+      return document;
+    });
     await this.client.bulk({ body: operations });
-    await this.client.indices.refresh({ index: this.indexName });
+
+    // aoss does not support refresh
+    if (!this.isAoss) {
+      await this.client.indices.refresh({ index: this.indexName });
+    }
```
Comment (huuyafwww) [inline libs/langchain-community/src/vectorstores/opensearch.ts:151]: #2302 to comply with the fixes.

File: libs/langchain-community/src/vectorstores/opensearch.ts [hunk 4]
```diff
@@ -256,10 +274,16 @@ export class OpenSearchVectorStore extends VectorStore {
...
+              match_mapping_type: "string",
               mapping: { type: "keyword" },
             },
           },
+          {
+              "metadata.loc": {
+                  match_mapping_type: "object",
+                  mapping: { type: "object" },
+              },
+          },
```
Comment (huuyafwww) [inline libs/langchain-community/src/vectorstores/opensearch.ts:286]: https://github.com/langchain-ai/langchainjs/issues/2036#issuecomment-1645769482 to comply with the fixes.

**Review by huuyafwww — COMMENTED** [review]
OpenSearchVectorStore now supports Amazon OpenSearch Serverless.
We would like to ask you to check this code.

File: libs/langchain-community/src/vectorstores/opensearch.ts [hunk 5]
```diff
@@ -256,10 +274,16 @@ export class OpenSearchVectorStore extends VectorStore {
           {
             // map all metadata properties to be keyword
             "metadata.*": {
-              match_mapping_type: "*",
+              match_mapping_type: "string",
```
Comment (jacoblee93) [inline libs/langchain-community/src/vectorstores/opensearch.ts:277]: Will this break anything?

**Comment by jacoblee93** [comment]
Small question, otherwise looks good!

**Comment by aqiao** [comment]
Hi guys,
Any updates on this, may i know the route map when the langchain-ai supports aoss?

**Comment by jacoblee93** [comment]
Had a small question above, also can you please run `yarn format` and  `yarn lint`?

**Comment by aqiao** [comment]
> Had a small question above, also can you please run `yarn format` and `yarn lint`?

I'm not familiar with NodeJS, just noticed its JS branch, so it's available for python branch already?

**Comment by jacoblee93** [comment]
Thank you!

**Pull request merged**